### PR TITLE
fix startswith/endswith implementation

### PR DESCRIPTION
**Pull request by bkamins**
This PR implements a check that prefix `b` does not match partial character in `a` in `startswith(a, b)`.

### CODE REVIEW

File: base/strings/util.jl [hunk 1]
```diff
@@ -45,9 +45,12 @@ function endswith(a::AbstractString, b::AbstractString)
 end
 endswith(str::AbstractString, chars::Chars) = !isempty(str) && last(str) in chars
 
-# FIXME: check that end of `b` doesn't match a partial character in `a`
-startswith(a::String, b::String) = sizeof(a) ≥ sizeof(b) &&
-    ccall(:memcmp, Int32, (Ptr{UInt8}, Ptr{UInt8}, UInt), a, b, sizeof(b)) == 0
+function startswith(a::String, b::String)
+    cub = ncodeunits(b)
+    (ncodeunits(a) < cub || ccall(:memcmp, Int32, (Ptr{UInt8}, Ptr{UInt8}, UInt),
+                                   a, b, sizeof(b)) ≠ 0) && return false
```
Comment (nalimilan) [inline base/strings/util.jl]: Missing space for alignment. Would probably be more readable as a `if`.

**Review by nalimilan — COMMENTED** [review]
@bkamins knows better than I do whether it's correct. For a serious review better ask @StefanKarpinski.

File: base/strings/util.jl (hunk 1 above)
Comment (bkamins) [reply base/strings/util.jl]: rewritten to use `if` now.

**Comment by bkamins** [comment]
The logic is that `b` is a prefix of `a` if next valid index in `a` is just after `ncodeunits(b)` codeunits in string `a`. The reason is that we know that in `b` position `ncodeunits(b)` is the end of a character in `b` (but it does not have to be the end in `a` in general so we have to test for this).

**Comment by vtjnash** [comment]
I don't understand the original TODO: what's wrong with doing the naive byte-comparison and declaring success if they match?

**Comment by bkamins** [comment]
I guess the original TODO says that it is not implemented yet 😄.
However given your question the answer is the following.
Consider strings:
```
x = "α"
y = "\xb1"
```
Now `endswith(x, y)` is `false` currently and it is a correct information.

However if you compared bytes you would say that there is a match because ` codeunit(x, 2)` is the same as `codeunit(y, 1)`.

**Comment by bkamins** [comment]
I have added `endswith` implementation and extended both to handle `SubString{String}`.

**Comment by bkamins** [comment]
Is this fix good to merge?

**Comment by StefanKarpinski** [comment]
It seems like way too much of a performance hit, although these particular functions are usually not performance-critical – https://github.com/JuliaLang/julia/pull/26921 is more of a problem. These both need the same fast primitive: a `memcmpx` function that returns the index of the first byte where they differ.

**Comment by bkamins** [comment]
Actually here the performance hit is much smaller than in #26921 and cannot be avoided (that is why here I wanted to merge, and I am waiting with the other one).
The reason is that where actually we can safely use `memcmp` (as I do now) and the only thing we have to check is if last characters in both strings are the same, which we have to do anyway (and `memcmpx` will not allow us to skip this test).

**Comment by StefanKarpinski** [comment]
Ok, fair enough. Let's merge this one then.

**Pull request merged**